### PR TITLE
Add validation that tests do not change global configs

### PR DESCRIPTION
**Pull request by jakevdp**

### CODE REVIEW

File: jax/_src/test_util.py [hunk 1]
```diff
@@ -1006,6 +1006,26 @@ def config_context(**kwds):
...
+  yield
+  ending_config = jax.config.values
+
+  if starting_config == ending_config:
+    return
+  differing = {k: (starting_config.get(k, NotPresent()), ending_config.get(k, NotPresent()))
+                for k in (starting_config.keys() | ending_config.keys())
+                if (k not in starting_config or k not in ending_config
+                    or starting_config[k] != ending_config[k])}
+  raise AssertionError(f"Test changed global config values. Differning values are: {differing}")
```
Comment (dfm) [inline jax/_src/test_util.py:1026]:
```suggestion
  raise AssertionError(f"Test changed global config values. Differing values are: {differing}")
```

Comment (jakevdp) [reply jax/_src/test_util.py]: Done, Thanks!

**Comment by jakevdp** [comment]
Compilation cache fix in https://github.com/google/jax/pull/21493

**Review by dfm — APPROVED** [review]
LGTM!

It looks like it needs a rebase after https://github.com/google/jax/commit/f04800f80da6c16ff6678d2f9cb6962b48329aae, but otherwise looks good.

We could add a simple test, for example testing that the following raises:

```python
with jtu.assert_global_configs_unchanged():
    jax.config.update("jax_enable_x64", not jax.config.values["jax_enable_x64"])
```

but I'm happy to skip that.

**Comment by jakevdp** [comment]
That test won't work unfortunately – it's not problematic to change configs during tests, it's only problematic if the original state is not restored, and the error is raised during `tearDown`. We could probably cover this by writing a test class that wraps `tearDown` in an `assertRaises`.

**Comment by dfm** [comment]
Right - I was actually suggesting something even simpler: Just testing that the context manager actually does raise on exit if the config changes. The larger question of how that fits into the absltest context manager is a bigger question, and that seemed beyond the scope here!

**Pull request merged**